specs/agents/tracing-spans-discarding.md

```diff
@@ -0,0 +1,128 @@
+should point to another valid transaction or span that's present in the Elasticsearch index.
+
+A span that refers to a missing span via is `parent_id` is also known as an "orphaned span".
+
+### Spans that propagate context to downstream services can't be discarded
+
+We only know whether to discard after the call has ended.
+At that point,
+the trace has already continued on the downstream service.
+Discarding the span for the external request would orphan the transaction of the downstream call.
+
+Propagating the trace context to downstream services is also known as out-of-process context propagation.
+
+### Intermediate spans
+
+Discarding a single span that has both a parent and a child span is not possible as it would lead to orphaned child spans.
+
+However,
+a whole subtree of spans can get discarded if all spans within that tree are requesting to be discarded.
+This means that if a leaf of the tree can't be discarded because it propagates context downstream,
+all spans leading up to it can't be discarded.
+If the leaf is a non-context propagating span,
+such as a manually created span or a SQL call,
+the subtree can be discarded.
+
+### Spans that lead up to async spans can't be discarded
+
+Whenever the context of a span is captured when scheduling an async operation,
+this span may not be discarded.
+
+That is because when the async operation is executed, 
+and the span context gets restored,
+the async operation might create child spans of the first span,
+even if it has already ended,
+or concurrently to it ending.
+In this scenario, discarding the first span would orphan the async span.
+
+This is even true if it's unknown whether the async operation creates a span at all.
+
+Propagating the trace context to async operations is also known as in-process context propagation.
+
+## Implementation
+
+Spans store two flags in order to determine whether a span can be discarded:
+- `discardable`:
+  Whether discarding this span is allowed.
+  The default value is `true`.
+  Setting this to `false` also sets the `discardable` flag of all it's parents and grand-parents to `false`.
+  See also [Intermediate spans](#Intermediate-spans).
+- `discardRequested`:
+  Whether this span is requested to be discarded.
+  The default value is `false`
+  
+### Marking as non-discardable
+
+According to the [limitations](#Limitations),
+there are certain situations where a span (and thus all it's parents) are marked as non-discardable.
+
+ - When an error is reported for this span
+ - When the span is reported to APM Server
+ - On out-of-process context propagation 
+ - On in-process context propagation
+
+### Request discarding
+
+If the span's duration is lower than `span_min_duration`,
+the span is requested to be discarded.
+
+### Determining whether to report a span
+
+If the span is requested to be discarded and discardable,
+the `span_count.dropped` count is incremeted and the span will not be reported.
```

Comment:
I assume that spans that are eventually dropped because of `span_min_duration` should not be counted in [`span_count.started`](https://github.com/elastic/apm-server/blob/7.0/docs/spec/transactions/transaction.json#L27). What implementation approach do you suggest?
1. Increment `span_count.started` at the start of each span and then decrement it for the spans that are dropped because of `span_min_duration`
2. Increment `span_count.started` at the end of each span when it's already known if the span is going to be dropped.

It seems that approach (1) has a disadvantage of potentially under-utilizing `transaction_max_spans` quota.

Reply:
All started spans count into that, no matter whether they are discarded. 
We don’t have to decrement when discarding as we also count discarded spans in another variable. See the section „ Impact on `transaction_max_spans`“ for more details.

Reply:
AFAIR spans dropped because of transaction_max_spans are not counted in span_count.started. Only spans actually sent to APM Server are counted in span_count.started. So there is no overlap between span_count.started and span_count.dropped.

Reply:
> Only spans actually sent to APM Server are counted in span_count.started.

That's news to me. Do you have any sources for that?

Reply:
Sorry, I was a bit confused it seems. You're right, the started count does not include dropped spans. The invariant `total = started + dropped` should always hold true. 

IMHO, `started` is a bit of a misnomer. Internally in the Java agent, the variable is named `reported`. It is incremented on span end if and only if the span is not discarded.
So it's the approach (2).
Note that the `started` count has no influence on `transaction_max_spans`. Only `total` and `dropped` is considered.

Reply:
I'm not sure what you mean by
> Note that the started count has no influence on transaction_max_spans. Only total and dropped is considered.

Isn't it the case that `started` should not be larger than `transaction_max_spans`?
If you increment `started` at the end of the span, you can get into situation where `started` is already equal to `transaction_max_spans` by the time the span ends so if you have to report the span (for example because its ID is referenced by an event that is already reported) you will have to increment `started` beyond `transaction_max_spans` - this is what I meant by
> over-utilizing `transaction_max_spans` quota

Reply:
the limit is reached if this condition is true:

    span_count.total - span_count.dropped >= transaction_max_spans

The `total` count is incremented for every span, even for those that are dropped (due to `transaction_max_spans` or `span_min_duration`).

This condition does not use the `started` count, hence it has no (direct) influence on `transaction_max_spans`.

Reply:
If
> The invariant `total = started + dropped` should always hold true.

and

> the limit is reached if this condition is true:
>
>    span_count.total - span_count.dropped >= transaction_max_spans

doesn't it follow that the limit is reached if this condition is true?

    span_count.started >= transaction_max_spans